### PR TITLE
[NTOS:PNP] Do not create an extra "FDO" device inside PnP root driver

### PR DESCRIPTION
**Pull request by Extravert-ir**
Implement the `PlugPlayControlInitializeDevice` control class, and cleanup the mess inside the PnP root driver.

JIRA Issue: [CORE-18989](https://jira.reactos.org/browse/CORE-18989)

### CODE REVIEW

**Review by binarymaster — COMMENTED** [review]
From build-linux (gcc, i386, Debug):
```
../src/ntoskrnl/io/pnpmgr/pnproot.c: In function 'PnpRootFdoPnpControl':
../src/ntoskrnl/io/pnpmgr/pnproot.c:992:35: error: variable 'DeviceExtension' set but not used [-Werror=unused-but-set-variable]
     PPNPROOT_FDO_DEVICE_EXTENSION DeviceExtension;
                                   ^~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

**Review by binarymaster — APPROVED** [review]
Tested, the bug is fixed, and I didn't noticed any new regressions :)

- ✅ **VBox BootCD with Standard HAL:** all 3 stages passed
- ✅ **VBox BootCD with ACPI HAL:** all 3 stages passed
- ✅ **VBox LiveCD with Standard HAL:** reboot & shutdown passing
- ✅ **VBox LiveCD with ACPI HAL:** reboot & shutdown passing

Additional tests:
- ✅ **Xbox port LiveCD:** reboot & shutdown passing
- ✅ **PC-98 port LiveCD:** reboot & shutdown passing

One suggestion though, you can add CORE-18989 to commit messages in all commits if you want.

**Comment by HBelusca** [comment]
@Extravert-ir Is it this PR that could fix the PS/2 keyboard & mouse not being usable on standard HAL ?

**Comment by HBelusca** [comment]
> > Is it this PR that could fix the PS/2 keyboard & mouse not being usable on standard HAL ?
> 
> I did tested the case of [CORE-17463](https://jira.reactos.org/browse/CORE-17463), it doesn't fix it.

Indeed; Victor confirmed the fix for PS/2 keyb&mouse is in a separate branch.

File: ntoskrnl/io/pnpmgr/pnproot.c [hunk 1]
```diff
@@ -38,39 +38,16 @@ typedef struct _PNPROOT_DEVICE
...
 /* Physical Device Object device extension for a child device */
 typedef struct _PNPROOT_PDO_DEVICE_EXTENSION
 {
-    // Common device data
-    PNPROOT_COMMON_DEVICE_EXTENSION Common;
     // Informations about the device
     PPNPROOT_DEVICE DeviceInfo;
 } PNPROOT_PDO_DEVICE_EXTENSION, *PPNPROOT_PDO_DEVICE_EXTENSION;
 
 /* Physical Device Object device extension for the Root bus device object */
```
[the remark this answers is not in the document]
Comment (Extravert-ir) [reply ntoskrnl/io/pnpmgr/pnproot.c:48]: That object is both FDO and PDO at the same time. Maybe I'll fix the comment with later refactorings

File: ntoskrnl/io/pnpmgr/pnproot.c [hunk 2]
```diff
@@ -181,28 +163,42 @@ PnpRootRegisterDevice(
     return STATUS_SUCCESS;
 }
 
+NTSTATUS
+PnpRootCreateDeviceObject(
+    OUT PDEVICE_OBJECT *DeviceObject)
+{
+    NTSTATUS status = IoCreateDevice(
```
[the remark this answers is not in the document]
Comment (Extravert-ir) [reply ntoskrnl/io/pnpmgr/pnproot.c:170]: I'll probably add more stuff here, so let's keep the current way for now

File: ntoskrnl/io/pnpmgr/pnproot.c [hunk 3]
```diff
@@ -1025,29 +1002,9 @@ PnpRootFdoPnpControl(
...
-
-         case IRP_MN_STOP_DEVICE:
-             DPRINT("IRP_MJ_PNP / IRP_MN_STOP_DEVICE\n");
-             /* Root device cannot be stopped */
-             Irp->IoStatus.Status = Status = STATUS_INVALID_DEVICE_REQUEST;
-             IoCompleteRequest(Irp, IO_NO_INCREMENT);
-             return Status;
-
         default:
+            // The root device object can receive only IRP_MN_QUERY_DEVICE_RELATIONS
```
[the remark this answers is not in the document]
Comment (Extravert-ir) [reply ntoskrnl/io/pnpmgr/pnproot.c:1006]: They are not filtered, the PnP manager should not send anything other than `IRP_MN_QUERY_DEVICE_RELATIONS`

File: ntoskrnl/io/pnpmgr/plugplay.c [hunk 4]
```diff
@@ -32,6 +32,9 @@ static KEVENT IopPnpNotifyEvent;
 
 /* FUNCTIONS *****************************************************************/
 
+NTSTATUS
+IopSetDeviceInstanceData(HANDLE InstanceKey, PDEVICE_NODE DeviceNode);
```
[the remark this answers is not in the document]
Comment (Extravert-ir) [reply ntoskrnl/io/pnpmgr/plugplay.c:36]: At some point, I will group all registry stuff in a separate file

**Comment by JoachimHenze** [comment]
Attention: It was reported that this PR did introduce regression:
[CORE-19353](https://jira.reactos.org/browse/CORE-19353) 'Regression: HWiNFO64 triggers an assertion'

**Pull request merged**